rfcs/2021-02-23-6531-performance-testing.md

```diff
@@ -0,0 +1,490 @@
+* We will encounter situations where benchmarks improve in CI and regress on
+  user's machines, especially as benchmarks become more "micro".
+
+### Build a `vector diagnostic` sub-command
+
+In this alternative we extend the Vector interface to include a `diagnostic`
+sub-command. This diagnostic will examine the system to gather information about
+it's running environment and perform, time fundamental actions. Information we
+might want to collect:
+
+* What operating system is Vector running on?
+* For the directories present in Vector's config, what filesystems are in use?
+  What mount options are in use?
+* How many CPUs are available to Vector and of what kind? How much memory?
+  NUMA?
+* What kernel parameters are set, especially those relating to common
+  bottlenecks like network, descriptor limits etc?
+* How long does a malloc/dealloc take for a series of block sizes?
+* How long does spawning a thread take?
+* How long do 2, 4, 8, 16 threads take to lock and unlock a common mutex?
+* For filesystems where Vector has R/W privileges, what IO characteristics
+  exist for these filesystems?
+
+This list is not exhaustive and hopefully you get the sense that the goal is to
+collect baseline information about the system to inform user issues and guide
+engineering work. An additional `doctor` sub-command could use the diagnostic
+feature to examine a config and make suggestions or point out easily detected
+issues, a disk buffer being configured to use a read-only filesystem, say.
+
+Does not address:
+
+* Regression from Increased Instrumentation
+* `metrics` Crate Upgrade Regresses Benchmarks
+* musl libc Release Builds
+* Channel Implementation Regresses Topology Throughput
+* Lua Transform Leaked Memory
+
+#### Upsides
+
+* We collect system information from users on a case by case basis. This
+  automates some of that information gathering.
+* Diagnostics information will help us discover unusual user systems that we
+  might otherwise struggle to reproduce.
+
+#### Downsides
+
+* The benefits of a `diagnostic` sub-command are not immediate, are focused on
+  the after-release side of regression and some users will not wish to share
+  its output with us.
+* A `diagnostic` sub-command is not a solution in itself but must be paired
+  with other approaches.
+
+### Run test-harness Nightly
+
+In this alternative we take the existing test-harness code base and adjust it to
+run nightly (likely UTC 00:00 for convenience), building from the current head
+of master branch. We will need to actually build a nightly Vector release for
+use by test-harness, but this seems straightforward to achieve.
+
+Addresses:
+
+* Regression from Increased Instrumentation
+* musl libc Release Builds
+* Channel Implementation Regresses Topology Throughput
+
+Does not address:
+
+* `metrics` Crate Upgrade Regresses Benchmarks
+* Lua Transform Leaked Memory
+
+#### Upsides
+
+* By adjusting the test-harness to run nightly we reduce the total number of
```

Comment:
I like this idea. My imagination is that "nightly" runs are achieved by plugging a SHA into a generic mechanism that sets up the tests and runs them with an eye toward doing this very kind of thing.

Reply:
That said, I see it as outside the scope of the RFC and more of a future piece of work, but I agree it would be _very_ useful, especially for exploration.

Reply:
I'll add a note in the RFC to say that we intend to preserve a stub for this behavior.

Reply:
Wording adjusted.